src/bindings/python/flux/job/executor.py

```diff
@@ -0,0 +1,529 @@
+            if self.__jobid is not None:
+                return self.__jobid
+            raise concurrent.futures.TimeoutError()
+
+    def add_jobid_callback(self, callback):
+        """Attaches a callable that will be called when the jobid is ready.
+
+        :param callback: A callable that will be called with this future as its only
+                argument when the future completes or is cancelled. The callable
+                will always be called by a thread in the same process in which
+                it was added. If the future has already completed or been
+                cancelled then the callable will be called immediately. These
+                callables are called in the order that they were added.
+        """
+        with self.__jobid_condition:
+            if self.__jobid is None:
+                self.__jobid_callbacks.append(callback)
+                return
+        self._invoke_flux_callback(callback)
+
+    def _invoke_flux_callback(self, callback, *args):
+        try:
+            callback(self, *args)
+        except Exception:  # pylint: disable=broad-except
+            logging.getLogger(__name__).exception(
+                "exception calling callback for %r", self
+            )
+
+    def exception(self, *args, **kwargs):
+        """If this method is invoked from a jobid callback by an executor thread,
+        it will result in deadlock, since the current thread will wait
+        for work that the same thread is meant to do.
+
+        Head off this possibility by checking the current thread.
+        """
+        if not self.done() and threading.get_ident() == self.__owning_thread_id:
+            raise RuntimeError("Cannot wait for future from inside callback")
+        return super().exception(*args, **kwargs)
+
+    exception.__doc__ = concurrent.futures.Future.exception.__doc__
+
+    def result(self, *args, **kwargs):
+        """If this method is invoked from a jobid callback by an executor thread,
+        it will result in deadlock, since the current thread will wait
+        for work that the same thread is meant to do.
+
+        Head off this possibility by checking the current thread.
+        """
+        if not self.done() and threading.get_ident() == self.__owning_thread_id:
+            raise RuntimeError("Cannot wait for future from inside callback")
+        return super().result(*args, **kwargs)
+
+    result.__doc__ = concurrent.futures.Future.result.__doc__
+
+
+# The FluxEventExecutorFuture has the same `.result()` and `.exception()` deadlock
+# issues as its superclass---but from event callbacks as well as jobid callbacks.
+class FluxEventExecutorFuture(FluxExecutorFuture):
+    """Subclass of ``FluxExecutorFuture`` providing additional job event updates.
+
+    Provides the ``add_event_callback`` method to invoke callbacks when particular
+    job-state events occur.
+
+    Valid events are contained in the ``EVENTS`` class attribute.
+    """
+
+    EVENTS = frozenset(
```

Comment:
I wonder if we should have these in a more "public" place so that other parts of the python bindings can leverage this (and the maintenance of this list is easier). Maybe in `flux.job.event`?  I was originally going to say we should avoid having these explicitly codified, but having the input validation on the event callback registration is quite nice.

Reply:
Yeah, I wanted to prevent users from adding callbacks that would never be called under any circumstances. `flux.job.event_wait` has a way of preventing bad event names, but only by waiting until the end of the event log. @grondo what do you think of adding `EVENTS` to `flux.job.event` and maybe adding a check `if name not in EVENTS: raise ValueError()` check to `flux.job.event_wait`?

Reply:
I wonder if the executor is kind of a special use case and so should expose a more limited set of events from the eventlog. (e.g. I doubt you need a "urgency" callback in this interface)

Reply:
Probably. I mean, I don't know who is going to use the executor's event callbacks other than Andre, Stephen for J/PSI, and maybe the Merlin team. I'm just going to guess that none of them will use the "urgency" event. But still---why restrict the events? If they don't want to know about the "urgency" event, they can just not add a callback for it. As long as the executor is listening for all events, I don't see why it shouldn't expose all of them to users.

Reply:
The only reason to restrict events is the maintenance issue alluded to by Stephen above. Also, I wrote that before the code could deal with multiple events. 

I guess from an "interface" point of view, I also think it is nicer when higher level interfaces can hide details, e.g. if FluxExecutor is being used mostly by workflow managers that really only care about when a job has been submitted (has a jobid), has started running, or has completed or hit an exception, it could be nice to present that level of interface instead of exposing the raw job eventlog. However, it may be too early for that anyway, and it was more of just a random thought thrown out there. I agree it is fine for the event watcher to return all events.

Reply:
>  @grondo what do you think of adding EVENTS to flux.job.event and maybe adding a check if name not in EVENTS: raise ValueError() check to flux.job.event_wait

I don't think we can do this. The interfaces in `flux.job.event` work for any eventlog not just the main eventlog and I think it would impractical (i.e. not possible) to list all possible events, especially when some of these eventlogs may take arbitrary events.

Reply:
of course, maybe we could do it only when `eventlog == "eventlog"`?

I like the idea of the master list of events in `flux.job.event`, so I'd personally be fine with moving it there if you'd like.

Reply:
Good points, thanks. I take it then that the main eventlog doesn't support arbitrary events, only the ones listed in RFC 21?

> I guess from an "interface" point of view, I also think it is nicer when higher level interfaces can hide details, e.g. if FluxExecutor is being used mostly by workflow managers that really only care about when a job has been submitted (has a jobid), has started running, or has completed or hit an exception

That is the only reason that `FluxExecutor` (as opposed to `FluxEventExecutor`) exists, aside from implementation and possible performance differences. `FluxEventExecutor` does everything that `FluxExecutor` does and more, but if you only want jobid/completion information (as Themis and Parsl do), the interface to `FluxExecutor is a little simpler.

Reply:
> I take it then that the main eventlog doesn't support arbitrary events, only the ones listed in RFC 21?

Hm, I don't think there is a hard rule there, but yes I think the intent is to document all possible events for the main job eventlog in RFC 21. (Nothing *prevents* a part of the system from appending a non-conforming event to the eventlog though)

Reply:
Sounds good. I moved the event list to `flux.job.event` but I didn't change anything else (like raising `ValueError`). I also made it so that if an event occurs that isn't in `EVENTS`, `FluxEventExecutor` will just ignore it, in the odd case that the system does append a non-conforming event.